File: app/src/main/java/com/stupidtree/hichat/data/source/BaseWebSource.java
package com.stupidtree.hichat.data.source;

import retrofit2.Retrofit;

public abstract class BaseWebSource<S> {
    Retrofit retrofit;
    protected S service;

    public BaseWebSource(Retrofit retrofit) {
        this.retrofit = retrofit;
        service = retrofit.create(getServiceClass());
    }

    protected abstract Class<S> getServiceClass();
}

File: app/src/main/java/com/stupidtree/cloudliter/data/source/BaseWebSource.kt
package com.stupidtree.cloudliter.data.source

import retrofit2.Retrofit

abstract class BaseWebSource<S>(retrofit: Retrofit) {
    @JvmField
    protected var service: S
    init {
        service = retrofit.create(getServiceClass())
    }

    protected abstract fun getServiceClass(): Class<S>


}